# Bioinformatic Algorithms Course Project - Dr. Koohi 
## Execution
`$ python prj.py [command] [options]`

## Commands
- BLK &#8594; blocking: will generate a file containing blocks. Options:
  + `--category` or `-c`:
    FLU for Influenza or MIT for mitochondrial.
  + `--sequence-index`  or  `-s`:
    file index in dataset.
  + `--block-size` or `-b`: 
    block size.
  + `--overlap` or `-o`:
    overlap between blocks.
  + `--filename` or `-f`:
    result file name contains blocks.

- CNT &#8594; counting kmers: will use a generated file of blocks and report each kmer occurrence and 8 bit scale. Options:
  + `--block-target` or `-t`:
    address of a file that contains blocks (generated by BLK command).
  + `-k`:
    kmer size.
  + `--filename` or `-f`:
    name for report file (kmers with their count and scale).
   
- CMP &#8594; comparing two sequence: printing (output stream) the Manhattan distance of two sequence. Options:
  +  `--category` or `-c`:
    FLU for Influenza or MIT for mitochondrial.
  +  `--sequence-index` or `-s`:
    file index of first sequence for comparing in dataset.
  +  `--other-index` or `-n`:
    file index of second sequence in dataset.
  +  `-k`:
    kmer size.
        
- CMP-B &#8594; comparing blocks of two sequences: will generate a file containing a distance matrix. Options:
  +  `--block-target` or `-t`:
    address of a fasta file contains blocks of first sequence.
  +  `--other-target` or `-h`:
    address of a fasta file contains blocks of second sequence.
  +  `-k`:
    kmer size.
  +  `--filename` or `-f`:
    result file name (contains the distance matrix).
        
- CMP-Z &#8594; comparing all sequences in same dataset (influenza and mitochondrial) - will generate a file with a matrix. Options:
  +  `--category` or `-c`:
    FLU for Influenza or MIT for mitochondrial.
  +  `-k`:
    kmer size.
  +  `--filename` or `-f`:
    result file name (contains the distance matrix).


Each option has its own default value. For more details, have a look at the source code.
